### PR TITLE
Add skip/src/dstnodata params to gdal2xyz

### PR DESCRIPTION
**Pull request by lpinner**
## Description

Adds `-skipnodata` and `-src/dstnodata` parameters to GDAL provider gdal2xyz processing algorithm. These parameters were added to GDAL's gdal2xyz.py utilitues as of GDAL 3.3, though the `-src/dstnodata` params are [broken as of GDAL 3.6](https://github.com/OSGeo/gdal/issues/7410).

Suggested help in [QGIS-Documentation gdal2xyz_params branch](https://github.com/lpinner/QGIS-Documentation/blob/gdal2xyz_params/docs/user_manual/processing_algs/gdal/rasterconversion.rst) (PR: [#8097](https://github.com/qgis/QGIS-Documentation/pull/8097))

### CODE REVIEW

File: python/plugins/processing/algs/gdal/gdal2xyz.py [hunk 1]
```diff
@@ -36,6 +37,11 @@
 class gdal2xyz(GdalAlgorithm):
     INPUT = 'INPUT'
     BAND = 'BAND'
+    if GdalUtils.version() > 3060200:
```
Comment (nyalldawson) [inline python/plugins/processing/algs/gdal/gdal2xyz.py:40]: The standard way to handle GDAL version dependent functionality is to ALWAYS add the parameters, but then raise a QgsProcessingException at runtime if the option is set but the GDAL version is too low. This approach means that users will get a user-friendly explanatory message as to why the parameter isn't available in their QGIS version (as opposed to just a missing option), and also ensures that models can be safely shared across different versions and that useful errors are raised if the operation of the model would vary depending on GDAL version...

Comment (lpinner) [reply python/plugins/processing/algs/gdal/gdal2xyz.py:40]: Thanks Nyall, good point, I'll address. 

I note that there's a version check in [gdalcalc.initAlgorithm`](https://github.com/qgis/QGIS/blob/master/python/plugins/processing/algs/gdal/gdalcalc.py#L148) so `self.addParameter` doesn't get called for `EXTENT` and `EXTENT_OPT` but no version check around the declaration of `EXTENT` and `EXTENT_OPT` class properties.  

As implemented this way, the options aren't shown in the Processing GUI but the parameters can still be passed in e.g. by a processing.run call resulting in the appropriate QgsProcessingException at runtime if GDAL version check fails.  

Is this the way you prefer it handled in this PR, or would you like the version check removed completely, except for the runtime check?

Comment (lpinner) [reply python/plugins/processing/algs/gdal/gdal2xyz.py]: @nyalldawson version checks removed except for runtime and tests.

File: python/plugins/processing/algs/gdal/gdal2xyz.py [hunk 2]
```diff
@@ -36,6 +37,9 @@
 class gdal2xyz(GdalAlgorithm):
     INPUT = 'INPUT'
     BAND = 'BAND'
+    SRCNODATA = 'SRCNODATA'
+    DSTNODATA = 'DSTNODATA'
+    SKIPNODATA = 'SKIPNODATA'
```
Comment (nyalldawson) [inline python/plugins/processing/algs/gdal/gdal2xyz.py:42]:
```suggestion
    SRCNODATA = 'NODATA_INPUT'
    DSTNODATA = 'NODATA_OUTPUT'
    SKIPNODATA = 'SKIP_NODATA'
```

That would match the naming used elsewhere in the gdal algorithms -- we tend toward uppercase, _ separated strings, where source = INPUT and dest = OUTPUT. See eg the merge algorithm.

File: python/plugins/processing/algs/gdal/gdal2xyz.py [hunk 3]
```diff
@@ -49,6 +53,16 @@ def initAlgorithm(self, config=None):
                                                      self.tr('Band number'),
                                                      1,
                                                      parentLayerParameterName=self.INPUT))
+
+        self.addParameter(QgsProcessingParameterNumber(self.SRCNODATA,
+                                                       self.tr('Use the given value as nodata'),
```
Comment (nyalldawson) [inline python/plugins/processing/algs/gdal/gdal2xyz.py:58]: Let's match these strings to the equivalents used by other algorithms:

```suggestion
                                                       self.tr('Input pixel value to treat as "nodata"'),
```

File: python/plugins/processing/algs/gdal/gdal2xyz.py [hunk 4]
```diff
@@ -49,6 +53,16 @@ def initAlgorithm(self, config=None):
...
+                                                       optional=True))  # GDAL > 3.6.3
+        self.addParameter(QgsProcessingParameterNumber(self.DSTNODATA,
+                                                       self.tr('Replace the nodata/srcnodata with the given nodata value.'),
```
Comment (nyalldawson) [inline python/plugins/processing/algs/gdal/gdal2xyz.py:61]:
```suggestion
                                                       self.tr('Assign specified "nodata" value to output'),
```

File: python/plugins/processing/algs/gdal/gdal2xyz.py [hunk 5]
```diff
@@ -49,6 +53,16 @@ def initAlgorithm(self, config=None):
...
+                                                       optional=True))  # GDAL > 3.6.3
+        self.addParameter(QgsProcessingParameterBoolean(self.SKIPNODATA,
+                                                        self.tr('Do not output nodata values.'),
```
Comment (nyalldawson) [inline python/plugins/processing/algs/gdal/gdal2xyz.py:64]:
```suggestion
                                                        self.tr('Do not output nodata values'),
```

**Comment by nyalldawson** [comment]
Just some minor wording changes left and then we're good to merge!

**Comment by lpinner** [comment]
> Just some minor wording changes left and then we're good to merge!

Thanks @nyalldawson, suggested changes incorporated and commits squashed.

File: python/plugins/processing/tests/GdalAlgorithmsRasterTest.py [hunk 6]
```diff
@@ -2533,7 +2533,47 @@ def testGdal2Xyz(self):
...
-                 outsource])
+                    outsource])
+
+            if GdalUtils.version() >= 3030000:
+                # skip nodata output
+                self.assertEqual(
+                    alg.getConsoleCommands({'INPUT': source,
+                                            'BAND': 1,
+                                            'CSV': False,
+                                            'SKIPNODATA': True,
```
Comment (nyalldawson) [inline python/plugins/processing/tests/GdalAlgorithmsRasterTest.py:2544]:
```suggestion
                                            'SKIP_NODATA': True,
```

File: python/plugins/processing/tests/GdalAlgorithmsRasterTest.py [hunk 7]
```diff
@@ -2533,7 +2533,47 @@ def testGdal2Xyz(self):
...
+                     outsource])
+
+            if GdalUtils.version() > 3060300:
+                # srcnodata output
+                self.assertEqual(
+                    alg.getConsoleCommands({'INPUT': source,
+                                            'BAND': 1,
+                                            'CSV': False,
+                                            'SRCNODATA': -999,
+                                            'SKIPNODATA': False,
```
Comment (nyalldawson) [inline python/plugins/processing/tests/GdalAlgorithmsRasterTest.py:2558]:
```suggestion
                                            'SKIP_NODATA': False,
```

File: python/plugins/processing/tests/GdalAlgorithmsRasterTest.py [hunk 8]
```diff
@@ -2533,7 +2533,47 @@ def testGdal2Xyz(self):
...
+                     source + ' ' +
+                     outsource])
+
+                # dstnodata output
+                self.assertEqual(
+                    alg.getConsoleCommands({'INPUT': source,
+                                            'BAND': 1,
+                                            'CSV': False,
+                                            'DSTNODATA': -999,
+                                            'SKIPNODATA': False,
```
Comment (nyalldawson) [inline python/plugins/processing/tests/GdalAlgorithmsRasterTest.py:2571]:
```suggestion
                                            'SKIP_NODATA': False,
```

File: python/plugins/processing/tests/GdalAlgorithmsRasterTest.py [hunk 9]
```diff
@@ -2533,7 +2533,47 @@ def testGdal2Xyz(self):
...
+                     source + ' ' +
+                     outsource])
+
+            if GdalUtils.version() > 3060300:
+                # srcnodata output
+                self.assertEqual(
+                    alg.getConsoleCommands({'INPUT': source,
+                                            'BAND': 1,
+                                            'CSV': False,
+                                            'SRCNODATA': -999,
```
Comment (nyalldawson) [inline python/plugins/processing/tests/GdalAlgorithmsRasterTest.py:2557]:
```suggestion
                                            'NODATA_INPUT': -999,
```

File: python/plugins/processing/tests/GdalAlgorithmsRasterTest.py [hunk 10]
```diff
@@ -2533,7 +2533,47 @@ def testGdal2Xyz(self):
...
+                     '-band 1 -srcnodata -999 ' +
+                     source + ' ' +
+                     outsource])
+
+                # dstnodata output
+                self.assertEqual(
+                    alg.getConsoleCommands({'INPUT': source,
+                                            'BAND': 1,
+                                            'CSV': False,
+                                            'DSTNODATA': -999,
```
Comment (nyalldawson) [inline python/plugins/processing/tests/GdalAlgorithmsRasterTest.py:2570]:
```suggestion
                                            'NODATA_OUTPUT': -999,
```

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**